### PR TITLE
Federated Read

### PR DESCRIPTION
**Pull request by Baunsgaard**
Read federated Matrices from file:

Layout:

```json
{
    "[0, 0] - [5, 13]": {
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X1",
        "dataType": "MATRIX",
        "address": "localhost:42441"
    },
    "[5, 0] - [10, 13]": {
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X2",
        "dataType": "MATRIX",
        "address": "localhost:36989"
    }
}
```

Metadata:
```json
{
    "data_type": "matrix",
    "value_type": "double",
    "rows": 10,
    "cols": 13,
    "nnz": -1,
    "format": "federated",
    "author": "baunsgaard",
    "created": "2020-10-25 18:55:02 CET"
}
```

### CODE REVIEW

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.sysds.common.Types;
+import org.apache.sysds.common.Types.DataType;
+import org.apache.sysds.conf.ConfigurationManager;
+import org.apache.sysds.runtime.DMLRuntimeException;
+import org.apache.sysds.runtime.io.IOUtilFunctions;
+import org.apache.sysds.runtime.meta.DataCharacteristics;
+
+public class ReaderFederated {
+    // private static final Log LOG = LogFactory.getLog(ReaderFederated.class.getName());
```
Comment (kev-inn) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java:50]: Remove?

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.sysds.common.Types;
+import org.apache.sysds.common.Types.DataType;
+import org.apache.sysds.conf.ConfigurationManager;
+import org.apache.sysds.runtime.DMLRuntimeException;
+import org.apache.sysds.runtime.io.IOUtilFunctions;
+import org.apache.sysds.runtime.meta.DataCharacteristics;
+
+public class ReaderFederated {
```
Comment (kev-inn) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java:49]: Move file to io?

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    private static long[] parseStringArray(String arr) {
+        String[] vv = arr.replaceAll("\\[", "").replaceAll("\\]", "").replaceAll("\\s", "").split(",");
+        long[] res = new long[vv.length];
+        for(int i = 0; i < vv.length; i++) {
+            res[i] = Long.parseLong(vv[i]);
+        }
+        return res;
+    }
```
Comment (kev-inn) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java:108]: I dislike the approach with ranges as json object names. Details below.

**Review by kev-inn — COMMENTED** [review]
In general I really like the PR, but I would propose one change for the json layout.
Current:
```json
{
    "[0, 0] - [5, 13]": {
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X1",
        "dataType": "MATRIX",
        "address": "localhost:42441"
    },
    "[5, 0] - [10, 13]": {
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X2",
        "dataType": "MATRIX",
        "address": "localhost:36989"
    }
}
```
My proposal:
```json
[
    {
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X1",
        "dataType": "MATRIX",
        "address": "localhost:42441",
        "begin": [0, 0],
        "end": [5, 13]
    },
    {
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X2",
        "dataType": "MATRIX",
        "address": "localhost:36989",
        "begin": [5, 0],
        "end": [10, 13]
    }
]
```

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java (hunk 2 above)
Comment (Baunsgaard) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java:49]: i was considering that, i am fine with either location

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java (hunk 1 above)
Comment (Baunsgaard) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java:50]: not this one. I want to keep them for easily adding logging for future.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java (hunk 3 above)
Comment (Baunsgaard) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/ReaderFederated.java:108]: me 2 do you have an alternative idea?

**Comment by Baunsgaard** [comment]
> In general I really like the PR, but I would propose one change for the json layout.
> Current:
> 
> ```json
> {
>     "[0, 0] - [5, 13]": {
>         "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X1",
>         "dataType": "MATRIX",
>         "address": "localhost:42441"
>     },
>     "[5, 0] - [10, 13]": {
>         "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X2",
>         "dataType": "MATRIX",
>         "address": "localhost:36989"
>     }
> }
> ```
> 
> My proposal:
> 
> ```json
> [
>     {
>         "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X1",
>         "dataType": "MATRIX",
>         "address": "localhost:42441",
>         "begin": [0, 0],
>         "end": [5, 13]
>     },
>     {
>         "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X2",
>         "dataType": "MATRIX",
>         "address": "localhost:36989",
>         "begin": [5, 0],
>         "end": [10, 13]
>     }
> ]
> ```

Sure will change it to that :) it looks nicer

**Comment by kev-inn** [comment]
I just checked the code again, wouldn't it be nicer to split up address and port too?

**Comment by Baunsgaard** [comment]
> I just checked the code again, wouldn't it be nicer to split up address and port too?

sure lets do that as well.
will look at it tomorrow morning

**Comment by Baunsgaard** [comment]
@kev-inn 

now it is: 

```json
[
    {
        "dataType": "MATRIX",
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X1",
        "begin": [
            0,
            0
        ],
        "end": [
            5,
            13
        ],
        "address": "localhost:41827"
    },
    {
        "dataType": "MATRIX",
        "filepath": "target/testTemp/functions/federated/io/FederatedReaderTest/in/X2",
        "begin": [
            5,
            0
        ],
        "end": [
            10,
            13
        ],
        "address": "localhost:41635"
    }
]

```

**Comment by kev-inn** [comment]
@Baunsgaard LGTM :+1:

**Pull request merged**